### PR TITLE
Swaparound Feature

### PR DESCRIPTION
**Pull request by mlysle**
This PR extends the swap feature with the optional ability to "swaparound" i.e. it can swap (<ins>x</ins>, y, z) -> (z, y, <ins>x</ins>). The option is disabled by default.

It might be hard to read the commits because my formatter changed all the spaces to tabs in my initial commits. Most of the actual changes happen in the left and right swap functions in swap.lua. In addition, I defined a new function in nodes.lua to find the "farthest" node.

### CODE REVIEW

**Comment by aaronik** [comment]
I see, ok thanks for the explanation.

The idea sounds cool. Honestly, the behavior of not swapping around was never explicitly defined. And the core concept doesn't seem particularly difficult to maintain. So honestly, I'd be down to just bring this in as core functionality - no config option at all. 

Howevery, it needs some things first:

1) Testing. There's a test suite in here that covers all the functionality of the plugin. No new functionality comes in until it's tested in many languages and edge cases are explored.
2) All the types are explicit (looking at farthest_sibling currently)
3) Commits get squashed

If all that stuff is included, and the config option is removed, then I'll be happy to merge this 🙂

**Comment by aaronik** [comment]
Awesome possum! Let me know if there's anything I can do to clarify.

Please note - this codebase is a codebase. And the tests are too. So, they're not perfect. At first I started grouping things together by action, like the highlight spec, and there was a movement spec. But it started getting so huge, I realized it probably should be broken up by the filetype it's testing. So forgive the leftovers!

File: tests/treewalker/c_sharp_spec.lua [hunk 1]
```diff
@@ -41,4 +41,12 @@ describe("In a C Sharp file", function()
     assert.same(first_block, lines.get_lines(30, 48))
     h.assert_cursor_at(7, 5)
   end)
+
+  it("swaparound behavior works", function()
+    vim.fn.cursor(52, 31) -- (|node1, node2)
```
Comment (aaronik) [inline tests/treewalker/c_sharp_spec.lua:46]: Can we remove all these -- (|node1, node2) comments where they're not accurate? Otherwise this is ready for merge!

File: tests/treewalker/haskell_spec.lua [hunk 2]
```diff
@@ -27,6 +28,12 @@ describe("In a haskell file: ", function()
...
   end)
-end)
-
 
+  it("swaparound behavior works", function()
+    vim.fn.cursor(40, 15)
+    tw.swap_left()
+    assert.same("  printEvens [9, 3, 5, 7, 1]", lines.get_line(40))
+    tw.swap_right()
+    assert.same("  printEvens [1, 3, 5, 7, 9]", lines.get_line(40))
```
Comment (aaronik) [inline tests/treewalker/haskell_spec.lua:37]: Hey so I pulled down your branch and tested out the functionality, and saw this test in particular. This is the only one with more than 2 parameters. At first I thought it was perfect but when I played with it, the swapping behavior isn't quite how I expect it to be. The way it is now the first and last items swap. This leads to these really interesting reorderings if you cycle through - a fascinating mathematical phenomenon, but not the desired behavior. What should happen is, if the node is at the end, it should just move itself to the other end. 

Here, use these for the haskell tests:

```lua
  it("swaparound behavior works", function()
    vim.fn.cursor(40, 15) -- [|1, 3, 5, 7, 9]
    assert.same("  printEvens [1, 3, 5, 7, 9]", lines.get_line(40))
    tw.swap_left()
    assert.same("  printEvens [3, 5, 7, 9, 1]", lines.get_line(40))
    tw.swap_left()
    assert.same("  printEvens [5, 7, 9, 1, 3]", lines.get_line(40))
    tw.swap_left()
    assert.same("  printEvens [7, 9, 1, 3, 5]", lines.get_line(40))
    tw.swap_left()
    assert.same("  printEvens [9, 1, 3, 5, 7]", lines.get_line(40))
    tw.swap_left() -- back to original order
    assert.same("  printEvens [1, 3, 5, 7, 9]", lines.get_line(40))

    vim.fn.cursor(40, 27) -- [|1, 3, 5, 7, |9]
    assert.same("  printEvens [1, 3, 5, 7, 9]", lines.get_line(40))
    tw.swap_right()
    assert.same("  printEvens [9, 1, 3, 5, 7]", lines.get_line(40))
    tw.swap_right()
    assert.same("  printEvens [7, 9, 1, 3, 5]", lines.get_line(40))
  end)

```

**Review by aaronik — COMMENTED** [review]
Was about to merge then caught one last thing, see the above comment. This is super close though!

**Comment by aaronik** [comment]
Hmmm that's a cool approach - but I worry about calling swap a bunch of times - that would overload users' undo list. Instead, maybe we can dispense with the idea of swapping two nodes entirely, and just remove the one node, then put it somewhere else? Is that feasible at all? Otherwise if that doesn't make sense, maybe your original solution is good, but with a bounds check, and if it's at the end, then it should just place the node at the beginning/end of the list? Maybe that's the same as the first suggestion..
